File: backend/src/main/java/com/ddfinance/backend/service/roles/EmployeeService.java
package com.ddfinance.backend.service.roles;

import com.ddfinance.backend.dto.actions.MessageDTO;
import com.ddfinance.backend.dto.investment.CreateInvestmentRequest;
import com.ddfinance.backend.dto.investment.InvestmentDTO;
import com.ddfinance.backend.dto.roles.ClientDTO;
import com.ddfinance.backend.dto.roles.EmployeeDetailsDTO;

import java.util.List;
import java.util.Map;

/**
 * Service interface for employee-specific operations.
 * Handles client management, investment operations, and employee metrics.
 *
 * @author Due Diligence Finance Team
 * @version 1.0
 * @since 2025-01-15
 */
public interface EmployeeService {

    /**
     * Gets employee details.
     *
     * @param email Employee's email
     * @return Employee details
     * @throws com.ddfinance.core.exception.EntityNotFoundException if employee not found
     */
    EmployeeDetailsDTO getEmployeeDetails(String email);

    /**
     * Gets all clients assigned to employee.
     *
     * @param email Employee's email
     * @return List of assigned clients
     * @throws com.ddfinance.core.exception.EntityNotFoundException if employee not found
     */
    List<ClientDTO> getAssignedClients(String email);

    /**
     * Gets specific client for employee.
     *
     * @param email Employee's email
     * @param clientId Client ID
     * @return Client details
     * @throws com.ddfinance.core.exception.EntityNotFoundException if not found
     * @throws com.ddfinance.core.exception.SecurityException.ForbiddenException if not assigned
     */
    ClientDTO getClientForEmployee(String email, Long clientId);

    /**
     * Gets investments for a client.
     *
     * @param email Employee's email
     * @param clientId Client ID
     * @return List of investments
     * @throws com.ddfinance.core.exception.SecurityException.ForbiddenException if not assigned
     */
    List<InvestmentDTO> getClientInvestments(String email, Long clientId);

    /**
     * Creates new investment for client.
     *
     * @param email Employee's email
     * @param request Investment creation request
     * @return Creation result with investment ID
     * @throws com.ddfinance.core.exception.ValidationException if invalid request
     * @throws com.ddfinance.core.exception.SecurityException.ForbiddenException if not assigned
     */
    Map<String, Object> createInvestment(String email, CreateInvestmentRequest request);

    /**
     * Updates investment status.
     *
     * @param email Employee's email
     * @param investmentId Investment ID
     * @param status New status
     * @return Updated investment
     * @throws com.ddfinance.core.exception.EntityNotFoundException if not found
     * @throws com.ddfinance.core.exception.ValidationException if invalid status
     */
    InvestmentDTO updateInvestmentStatus(String email, Long investmentId, String status);

    /**
     * Sends message to client.
     *
     * @param email Employee's email
     * @param clientId Client ID
     * @param message Message to send
     * @return Send result
     * @throws com.ddfinance.core.exception.SecurityException.ForbiddenException if not assigned
     */
    Map<String, Object> sendMessageToClient(String email, Long clientId, MessageDTO message);

    /**
     * Gets messages from clients.
     *
     * @param email Employee's email
     * @return List of messages
     */
    List<Map<String, Object>> getMessagesFromClients(String email);

    /**
     * Gets employee performance metrics.
     *
     * @param email Employee's email
     * @return Performance metrics map
     */
    Map<String, Object> getPerformanceMetrics(String email);

    /**
     * Searches assigned clients.
     *
     * @param email Employee's email
     * @param query Search query
     * @return Matching clients
     */
    List<ClientDTO> searchClients(String email, String query);

    /**
     * Generates client report.
     *
     * @param email Employee's email
     * @param clientId Client ID
     * @param format Report format
     * @return Report data
     * @throws com.ddfinance.core.exception.SecurityException.ForbiddenException if not assigned
     */
    byte[] generateClientReport(String email, Long clientId, String format);

    /**
     * Gets pending investments.
     *
     * @param email Employee's email
     * @return List of pending investments
     */
    List<InvestmentDTO> getPendingInvestments(String email);

    /**
     * Updates client notes.
     *
     * @param email Employee's email
     * @param clientId Client ID
     * @param notes Updated notes
     * @throws com.ddfinance.core.exception.SecurityException.ForbiddenException if not assigned
     */
    void updateClientNotes(String email, Long clientId, String notes);

    /**
     * Gets employee calendar/schedule.
     *
     * @param email Employee's email
     * @return Calendar events
     */
    List<Map<String, Object>> getEmployeeSchedule(String email);

    /**
     * Records client meeting.
     *
     * @param email Employee's email
     * @param clientId Client ID
     * @param meetingDetails Meeting details
     * @return Meeting record ID
     */
    Long recordClientMeeting(String email, Long clientId, Map<String, Object> meetingDetails);
}

File: frontend/src/api/EmployeeService.ts
// frontend/src/api/EmployeeService.ts

import { apiClient } from './apiClient';
import {
  ClientDTO,
  InvestmentDTO,
  CreateInvestmentRequest,
  UpdateInvestmentRequest,
  CreateMessageRequest,
  MessageDTO,
  PageRequest,
  Page,
  EmployeeDTO,
} from '@/types';

/**
 * Service for handling all employee-specific API operations.
 * Includes client management, investment processing, and communication.
 */
export class EmployeeService {
  private static instance: EmployeeService;

  private constructor() {
    // Private constructor to enforce singleton pattern
  }

  /**
   * Get the singleton instance of EmployeeService
   */
  public static getInstance(): EmployeeService {
    if (!EmployeeService.instance) {
      EmployeeService.instance = new EmployeeService();
    }
    return EmployeeService.instance;
  }

  // Client Management
  // =================

  /**
   * Get paginated list of assigned clients
   * @param pageRequest - Pagination parameters
   * @returns Page of assigned clients
   */
  public async getAssignedClients(pageRequest?: PageRequest): Promise<Page<ClientDTO>> {
    const response = await apiClient.get<Page<ClientDTO>>('/employees/clients', {
      params: pageRequest,
    });
    return response.data;
  }

  /**
   * Get detailed information for a specific client
   * @param clientId - Client ID
   * @returns Client details
   */
  public async getClient(clientId: number): Promise<ClientDTO> {
    const response = await apiClient.get<ClientDTO>(`/employees/clients/${clientId}`);
    return response.data;
  }

  /**
   * Search for clients by name, email, or other criteria
   * @param query - Search query
   * @returns List of matching clients
   */
  public async searchClients(query: string): Promise<ClientDTO[]> {
    const response = await apiClient.get<ClientDTO[]>('/employees/clients/search', {
      params: { q: query },
    });
    return response.data;
  }

  // Investment Management
  // ====================

  /**
   * Get all investments for a specific client
   * @param clientId - Client ID
   * @returns List of client investments
   */
  public async getClientInvestments(clientId: number): Promise<InvestmentDTO[]> {
    const response = await apiClient.get<InvestmentDTO[]>(
      `/employees/clients/${clientId}/investments`
    );
    return response.data;
  }

  /**
   * Create a new investment for a client
   * @param clientId - Client ID
   * @param request - Investment creation request
   * @returns Created investment
   */
  public async createInvestmentForClient(
    clientId: number,
    request: CreateInvestmentRequest
  ): Promise<InvestmentDTO> {
    const response = await apiClient.post<InvestmentDTO>(
      `/employees/clients/${clientId}/investments`,
      request
    );
    return response.data;
  }

  /**
   * Update a client's investment
   * @param clientId - Client ID
   * @param investmentId - Investment ID
   * @param updateRequest - Update data
   * @returns Updated investment
   */
  public async updateClientInvestment(
    clientId: number,
    investmentId: number,
    updateRequest: UpdateInvestmentRequest
  ): Promise<InvestmentDTO> {
    const response = await apiClient.put<InvestmentDTO>(
      `/employees/clients/${clientId}/investments/${investmentId}`,
      updateRequest
    );
    return response.data;
  }

  /**
   * Process a pending investment order
   * @param orderId - Order ID
   * @param action - Action to take (APPROVE, REJECT, etc.)
   * @returns Processing result
   */
  public async processInvestmentOrder(orderId: number, action: string): Promise<any> {
    const response = await apiClient.post(`/employees/orders/${orderId}/process`, { action });
    return response.data;
  }

  // Dashboard & Analytics
  // ====================

  /**
   * Get employee dashboard statistics
   * @returns Dashboard stats including clients, AUM, pending orders
   */
  public async getDashboardStats(): Promise<any> {
    const response = await apiClient.get('/employees/dashboard/stats');
    return response.data;
  }

  /**
   * Get all pending investment orders requiring approval
   * @returns List of pending orders
   */
  public async getPendingOrders(): Promise<any[]> {
    const response = await apiClient.get('/employees/orders/pending');
    return response.data;
  }

  /**
   * Get employee performance metrics
   * @param period - Time period (monthly, quarterly, yearly)
   * @returns Performance metrics
   */
  public async getPerformanceMetrics(period: string = 'monthly'): Promise<any> {
    const response = await apiClient.get('/employees/performance', {
      params: { period },
    });
    return response.data;
  }

  // Communication
  // =============

  /**
   * Send a message to a client
   * @param clientId - Client ID
   * @param message - Message content
   * @returns Sent message
   */
  public async sendMessageToClient(
    clientId: number,
    message: CreateMessageRequest
  ): Promise<MessageDTO> {
    const response = await apiClient.post<MessageDTO>(
      `/employees/clients/${clientId}/messages`,
      message
    );
    return response.data;
  }

  /**
   * Get all messages with a specific client
   * @param clientId - Client ID
   * @returns List of messages
   */
  public async getClientMessages(clientId: number): Promise<MessageDTO[]> {
    const response = await apiClient.get<MessageDTO[]>(`/employees/clients/${clientId}/messages`);
    return response.data;
  }

  /**
   * Get all unread messages across all clients
   * @returns List of unread messages
   */
  public async getUnreadMessages(): Promise<MessageDTO[]> {
    const response = await apiClient.get<MessageDTO[]>('/employees/messages/unread');
    return response.data;
  }

  // Employee Profile
  // ================

  /**
   * Get current employee's profile
   * @returns Employee profile
   */
  public async getProfile(): Promise<EmployeeDTO> {
    const response = await apiClient.get<EmployeeDTO>('/employees/me/profile');
    return response.data;
  }

  /**
   * Update employee profile
   * @param profileData - Profile data to update
   * @returns Updated profile
   */
  public async updateProfile(profileData: Partial<EmployeeDTO>): Promise<EmployeeDTO> {
    const response = await apiClient.put<EmployeeDTO>('/employees/me/profile', profileData);
    return response.data;
  }

  // Appointments & Schedule
  // ======================

  /**
   * Get employee's appointments
   * @param startDate - Start date for appointments
   * @param endDate - End date for appointments
   * @returns List of appointments
   */
  public async getAppointments(startDate?: string, endDate?: string): Promise<any[]> {
    const response = await apiClient.get('/employees/appointments', {
      params: { startDate, endDate },
    });
    return response.data;
  }

  /**
   * Schedule a new appointment with a client
   * @param appointment - Appointment details
   * @returns Created appointment
   */
  public async scheduleAppointment(appointment: any): Promise<any> {
    const response = await apiClient.post('/employees/appointments', appointment);
    return response.data;
  }

  // Reports & Analytics
  // ==================

  /**
   * Generate client portfolio report
   * @param clientId - Client ID
   * @param reportType - Type of report
   * @returns Report data or URL
   */
  public async generateClientReport(clientId: number, reportType: string): Promise<any> {
    const response = await apiClient.post(`/employees/clients/${clientId}/reports`, {
      type: reportType,
    });
    return response.data;
  }

  /**
   * Get market analysis and insights
   * @returns Market analysis data
   */
  public async getMarketAnalysis(): Promise<any> {
    const response = await apiClient.get('/employees/market-analysis');
    return response.data;
  }
}

// Export singleton instance
export const employeeService = EmployeeService.getInstance();
